lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java

```diff
@@ -151,6 +165,78 @@ public boolean matches(DictionaryDescriptor dictionary)
         return effectivePredicateDomain == null || effectivePredicateMatches(effectivePredicateDomain, dictionary);
     }
 
+    private Optional<Collection<Object>> extractDiscreteValues(ValueSet valueSet)
+    {
+        if (!valueSet.isDiscreteSet()) {
+            return valueSet.tryExpandRanges(domainCompactionThreshold);
+        }
+
+        return Optional.of(valueSet.getDiscreteSet());
+    }
+
+    @Override
+    public boolean matches(ColumnDescriptor columnDescriptor, ColumnPath columnPath, BloomFilterStore bloomFilterStore)
+    {
+        requireNonNull(bloomFilterStore, "bloomFilterStore is null");
+
+        if (effectivePredicate.isNone()) {
+            return false;
+        }
+        Map<ColumnDescriptor, Domain> effectivePredicateDomains = effectivePredicate.getDomains()
+                .orElseThrow(() -> new IllegalStateException("Effective predicate other than none should have domains"));
+
+        if (!effectivePredicateDomains.containsKey(columnDescriptor)) {
+            return true;
+        }
+
+        Domain effectivePredicateDomain = effectivePredicateDomains.get(columnDescriptor);
+        // the bloom filter bitset contains only non-null values so isn't helpful
+        if (effectivePredicateDomain.isNullAllowed()) {
+            return true;
+        }
+        Optional<BloomFilter> bloomFilterOptional = bloomFilterStore.readBloomFilter(columnPath);
+        if (!bloomFilterOptional.isPresent()) {
+            return true;
+        }
+
+        Optional<Collection<Object>> discreteValues = extractDiscreteValues(effectivePredicateDomain.getValues());
+        if (discreteValues.isEmpty()) {
```

Comment:
Would it be better to do this before reading the bloom filter?. if the bloom filter is not useful then we can avoid reading it.

Reply:
thanks, updated.

Reply:
made sure bloom filter is only read at the very end when necessary.